site-src/geps/gep-724.md

```diff
@@ -0,0 +1,614 @@
+
+#### 6. We Need to Maintain a Handshake Between Gateways and Routes
+Of course we do still need a handshake that will enable cross-namespace
+references between Routes and Gateways. This proposal leaves in the core
+capabilities of the v1alpha1 API for this. Gateways can specify the namespaces
+they trust Routes to bind from, and Routes directly reference the Gateways they
+want to attach to. This is largely similar to the ReferencePolicy model proposed
+for Route->Service references, but is embedded within the Route and Gateway
+resources. The alternatives below explore what this could look like with
+ReferencePolicy.
+
+## API Changes
+
+The proposed changes here can be summarized as:
+
+* Remove Route selector from the RouteBindingSelector in Gateway listeners.
+* Replace Route kind and group with optional list of accepted kinds and groups
+  in RouteBindingSelector.
+* Rename RouteBindingSelector to ListenerRoutes.
+* Replace the 3 options from Route -> Gateway (All, FromList, SameNamespace)
+  with a reference list that supports arbitrary kinds.
+
+### Gateway
+
+On Gateway, the only change involves removing the Route selector field.
+Everything else remains the same.
+
+#### Removed
+
+```go
+  // Selector specifies a set of route labels used for selecting
+  // routes to associate with the Gateway. If this Selector is defined,
+  // only routes matching the Selector are associated with the Gateway.
+  // An empty Selector matches all routes.
+  //
+  // Support: Core
+  //
+  // +optional
+  Selector *metav1.LabelSelector `json:"selector,omitempty"`
+	// Group is the group of the route resource to select. Omitting the value or specifying
+	// the empty string indicates the networking.x-k8s.io API group.
+	// For example, use the following to select an HTTPRoute:
+	//
+	// routes:
+	//   kind: HTTPRoute
+	//
+	// Otherwise, if an alternative API group is desired, specify the desired
+	// group:
+	//
+	// routes:
+	//   group: acme.io
+	//   kind: FooRoute
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:default=networking.x-k8s.io
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Group *string `json:"group,omitempty"`
+	// Kind is the kind of the route resource to select.
+	//
+	// Kind MUST correspond to kinds of routes that are compatible with the
+	// application protocol specified in the Listener's Protocol field.
+	//
+	// If an implementation does not support or recognize this
+	// resource type, it SHOULD set the "ResolvedRefs" condition to false for
+	// this listener with the "InvalidRoutesRef" reason.
+	//
+	// Support: Core
+	Kind string `json:"kind"`
+```
+
+#### Added
+```go
+type ListenerRoutes struct {
+  // ...
+  // Kinds specifies the groups and kinds of Routes that are allowed to bind to
+  // this Gateway listener. When unspecified or empty, the only limitation on
+  // the kinds of Routes supported is the Listener protocol. Kind MUST
```

Comment:
It would be better to have an explicit list over an implied reference. The map of protocol->[]route-types-on-that-protocol will change based on the implementations and even within core as time goes on, which could lead to unnecessary exposure to other protocols.

If YAML verbosity is the reason, IMHO we should pick clarity in this case over conciseness.
Another solution would be to introduce defaulting via the webhook (need a mutating one) based on the listener protocol.

Reply:
I disagree. Not sure I can provide any concrete reason why other than personal preference though :slightly_smiling_face:.

It seems to me there is minimal risk of accidentally selecting the wrong kind - and even if it does happen, is it so bad? I am assuming we don't add an InsecureHTTPRoute that violates a bunch of security rules in the future (or a less contrived example) where it would be bad to select extra routes.

Reply:
I'm not tied to either of these approaches, but do prefer the more lenient interpretation I proposed due to ease of use. With the change to direct references proposed here, the risk of routes being accidentally bound to Gateways seems very low.

If we were still doing selection, a change to the Routes supported for a given protocol could immediately result in more Routes being attached to a Gateway when Route types were not specified. On the other hand, with direct references from Routes, a Route of the new type would have to directly reference the Gateway it wanted to attach to. That seems unlikely to cause things to break in significant ways. 

I'd completely agree that we should avoid this if we were still using selection, but with this new approach, I think the only thing most Gateway admins should have to worry about is the hostname, port, and protocol of a listener + the namespaces it supports. If you need to further restrict Route kinds, that's possible, but that seems like an advanced and usually unnecessary use case.

Reply:
> If we were still doing selection, a change to the Routes supported for a given protocol could immediately result in more Routes being attached to a Gateway when Route types were not specified. On the other hand, with direct references from Routes, a Route of the new type would have to directly reference the Gateway it wanted to attach to. That seems unlikely to cause things to break in significant ways.

Fair. I'm sold.

How about having defaults in here?
If I'm debugging or trying to understand how the API works, and I see '[]RouteGroupKind' as empty, how I'm supposed to know which types routes are binding to this listener since there is no concept of 'protocol' in Routes.
I could do filtering and what not but seeing the route types in there will give a clear pointer to look at those other resources.

Reply:
> I'd completely agree that we should avoid this if we were still using selection, but with this new approach, I think the only thing most Gateway admins should have to worry about is the hostname, port, and protocol of a listener + the namespaces it supports. If you need to further restrict Route kinds, that's possible, but that seems like an advanced and usually unnecessary use case.

I think this is the argument that clinches this for me. As long as we mandate what controllers must do for the status of an object that attempts to attach to a listener that can't support it, I think it's okay.

However, I do think that it's important to have some status information that makes it clear to a Gateway owner that Routes have attempted to attach to a Listener that doesn't support them - we need to ensure there's a trail in both Gateway -> Route and Route -> Gateway directions for people to follow when things go wrong, because the objects have different owners.

In terms of having defaults, I think that it's important for the described behavior that this be defaulted to an empty list, but we can specify defaults in the RouteGroupKind struct. I feel like @hbagdi's concern here is about discoverability and status though?

Reply:
Thanks @youngnick, I think that's a good distinction. We could expand status for Gateway listeners so controllers could populate the kinds of Routes they were supporting for that listener. I've got some larger thoughts around status that will require a bigger update, so I'll work on that in a subsequent commit. @hbagdi would a status based approach work for you in terms of improving debuggability?

Reply:
Added a new status based approach as described above.